sips/sip-019/sip-019-token-metadata-update-notifications.md

```diff
@@ -0,0 +1,309 @@
+parse and validate its contents, and refresh any contracts that were updated.
+
+# Specification
+
+Notification messages for each token class are specified below. Token metadata update notifications
+must be made via a contract call transaction to the [deployed reference
+contract](https://explorer.stacks.co/txid/0xe92af2ea5c11e2e6fde4d31fd394de888070efff23bffad04465c549543abfa2?chain=mainnet)
+or from a call to `print` within any other contract, including the token contract itself.
+
+## Fungible Tokens
+
+When a contract needs to notify the network that metadata has changed for a **Fungible Token**, it
+shall call `print` with a tuple with the following structure:
+
+```clarity
+{ notification: "token-metadata-update", payload: { token-class: "ft", contract-id: <token contract id> }}
+```
+
+| Key                   | Value                                                                  |
+|-----------------------|------------------------------------------------------------------------|
+| `notification`        | The string `"token-metadata-update"`                                   |
+| `payload.token-class` | The string `"ft"`                                                      |
+| `payload.contract-id` | The contract id (principal) of the contract that declared the token    |
+
+## Non-Fungible Tokens
+
+When a contract needs to notify the network that metadata has changed for a **Non-Fungible Token**,
+it shall call `print` with a tuple with the following structure:
+
+```clarity
+{ notification: "token-metadata-update", payload: { token-class: "nft", token-ids: (list u100, u101), contract-id: <token contract id> }}
+```
+
+| Key                   | Value                                                                  |
+|-----------------------|------------------------------------------------------------------------|
+| `notification`        | The string `"token-metadata-update"`                                   |
+| `payload.token-class` | The string `"nft"`                                                     |
+| `payload.token-ids`   | (optional) A list with the uint token ids that need to be refreshed    |
+| `payload.contract-id` | The contract id (principal) of the contract that declared the tokens   |
+
+If a notification does not contain a value for `payload.token-ids`, it means it is requesting an
+update for all tokens.
+
+## Considerations for metadata indexers
+
+For a token metadata update notification to be considered valid by metadata indexers, it must meet
+the following requirements:
+
+1. Its payload structure should be correct whether it is updating a [FT](#fungible-tokens) or a
+   [NFT](#non-fungible-tokens).
+1. The STX address that initiated the contract call which emits the notification should match the
+   owner of the token contract being updated.
+    * The transaction's `tx-sender` principal should match the principal contained in the
+      notification's `payload.contract-id`.
```

Comment:
This seems like an overly aggressive permissions check -- it means that token metadata can only ever be updated by exactly the address that published the contract. Some tokens may use a different "controller" principal from the publisher. An alternative permissions check could be:

"The `contract_identifier` field of the contract event must be equal to the `payload.contract-id`"

This means that the token metadata update event was emitted by the correct contract, allowing the contract itself to perform any permissions checks that it wants for metadata update notifications.

Reply:
> "The contract_identifier field of the contract event must be equal to the payload.contract-id"

@kantai if I understand correctly, this check would mean that only a contract would be able to update its own tokens, correct?

By doing that, though, we'd lose backwards compatibility with contracts that are already deployed and want to update their tokens retroactively.

Reply:
That sounds great @kantai , updating now...